### DISCUSSION TITLE
What is the status of GraphQL-Java Annotations support?

### DISCUSSION THREAD

**Discussion by devopsix** [graphql-java-kickstart/graphql-spring-boot#500] [Q&A]
EDIT: Renamed from ~“Where to find the example project with the GraphQL Annotations Spring DI work-arounds?”~ to “What is the status of GraphQL-Java Annotations support?”

Support for GraphQL-Java Annotations seems to be disabled by default. Its is only enabled after including the following dependency:

```xml
<dependency>
    <groupId>com.graphql-java-kickstart</groupId>
    <artifactId>graphql-kickstart-spring-boot-starter-graphql-annotations</artifactId>
    <version>8.1.1</version>
</dependency>
```

In the Spring Boot world, I would rather have expected that support is enabled by just adding the GraphQL-Java Annotations dependency:

```xml
<dependency>
    <groupId>io.github.graphql-java</groupId>
    <artifactId>graphql-java-annotations</artifactId>
    <version>8.3</version>
</dependency>
```

From the [Root resolvers, directives, type extensions](https://github.com/graphql-java-kickstart/graphql-spring-boot#root-resolvers-directives-type-extensions) section of the graphql-spring-boot's README regarding the GraphQL-Java Annotations library:

> This means that Spring dependency injection will not work in the usual way. The companion example project (which can be found in the [samples](https://github.com/graphql-java-kickstart/samples) repository) demonstrates possible workarounds for this issue.

I have searched the samples repository for “annotations”, “@GraphQLField” and “@GraphQLQueryResolver” but was unable to find any GraphQL-Java Annotations example at all.

I noticed that a GraphQL-Java Annotations example was added to the graphql-spring-boot repository in c8f0a724698f9b17913cb637c769f23f6616933f but was instantly removed in d37d5b931e778a05efac5aba26dd84717f8f2f80. [`Mutations.java`](example-graphql-annotations/src/main/java/graphql/kickstart/spring/boot/graphql/annotations/example/resolvers/Mutations.java) might be one of the examples I was looking for.

That's giving me a surprisingly hard time using a feature advertized on the project's home page. I am wondering what the status of GraphQL-Java Annotations support actually is.

Maintainers' feedback on this would be highly appreciated. Can GraphQL-Java Annotations support be considered stable? Are any contributions needed?

**Comment by oliemansm** [comment]
Contributions are definitely welcome. The way the project has been setup at the moment is with a starter per functionality as you have noticed. It's worth considering to alter that to the approach you mention with one starter and adding a specific dependency to activate certain functionality. Although Spring Boot also uses both approaches. If you want to use `web` or `webflux` you have to choose a particular starter. Same applies here if you want to use graphql-java-tools you'll need a different starter than when you want to use annotations.

Support for graphql-java-annotations has been added by one of our active contributors @BlasiusSecundus. The sample was moved from this project to our samples project, however that PR is still a draft because of some weird issue, see: https://github.com/graphql-java-kickstart/samples/pull/7.

As stated on our homepage we are looking for contributors, so any help is appreciated.

**Reply by devopsix** [reply]
Thank you very much for your prompt response!

> Same applies here if you want to use graphql-java-tools you'll need a different starter than when you want to use annotations.

I think there are some changes that would have helped me a lot understanding what approach you have chosen:
- Include `tools` in the name of the dedicated GraphQL Java Tools starter. `graphql-spring-boot-starter` is rather unspecific compared to `graphql-kickstart-spring-boot-starter-graphql-annotations`.
- Use consistent and brief naming for the “public” starter dependencies. `graphql-kickstart-spring-boot-starter-graphql-annotations` sounded “internal” to me.
- ~Make the library specific starters include the actual library dependency.~ EDIT: That already is the case.
- Describe in the README what starter to use for what library. (PR [documentation#9](https://github.com/graphql-java-kickstart/documentation/pull/9))
- Mention the annotations starter in the [GraphQL Spring Boot documentation](https://www.graphql-java-kickstart.com/spring-boot/)  (PR #503)

**Comment by BlasiusSecundus** [comment]
I have somewhat neglected the sample project lately, but I plan to update it shortly using the latest release (currently uses an old snapshot) of the starter and finalize the PR.

As for the stability / productin readiness: it is certainly less tested and used than the graphql-tools approach. Some major shortcomings (like the issues with dependency injection) need improvements in the GraphQL Annotations depencency first.

Update:
Also, the annotations starter currently only works with the `web` starter (which it includes as api dependency), so it needs some improvements to make it work with `webflux` too.

Update 2:

That strange test failure still persists. I will continue investigating and try to find the cause.

**Reply by devopsix** [reply]
@BlasiusSecundus, have you already looked into [graphql-java-annotations#262](https://github.com/Enigmatis/graphql-java-annotations/issues/262)? Would upgrading to graphql-java-annotations:8.3 and using `@AllArgsConstructor(onConstructor_ = @GraphQLConstructor)` on `CreatePerson` maybe fix the issue?

**Reply by BlasiusSecundus** [reply]
Thanks for the idea, I will try that.

**Reply by BlasiusSecundus** [reply]
It works, thanks!
